dra: pre-scheduled pods

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

When someone decides that a Pod should definitely run on a specific node, they
can create the Pod with spec.nodeName already set. Some custom scheduler might
do that. Then kubelet starts to check the pod and (if DRA is enabled) will
refuse to run it, either because the claims are still waiting for the first
consumer or the pod wasn't added to reservedFor. Both are things the scheduler
normally does.
    
Also, if a pod got scheduled while the DRA feature was off in the
kube-scheduler, a pod can reach the same state.
    
#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/114005

#### Special notes for your reviewer:

The resource claim controller can handle these two cases by taking over for the
kube-scheduler when nodeName is set. Triggering an allocation is simpler than
in the scheduler because all it takes is creating the right
PodSchedulingContext with spec.selectedNode set. There's no need to list nodes
because that choice was already made, permanently. Adding the pod to
reservedFor also isn't hard.
    
What's currently missing is triggering de-allocation of claims to re-allocate
them for the desired node. This is not important for claims that get created
for the pod from a template and then only get used once, but it might be
worthwhile to add de-allocation in the future.


#### Does this PR introduce a user-facing change?
```release-note
kube-controller-manager: the dynamic resource controller steps in when a pod got created such that the scheduler ignores it (i.e. spec.nodeName is set) and then takes care of triggering delayed resource claim allocation and/or reserving a claim for the pod.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
-KEP: https://github.com/kubernetes/enhancements/issues/3063
```

pohly committed:
resourceclaim controller: avoid caching deleted pod unnecessarily

We don't need to remember that a pod got deleted when it had no resource claims
because the code which checks the cached UIDs only checks for pods which have
resource claims.

pohly committed:
dra: test with and without immediate ReservedFor

The recommendation and default in the controller helper code is to set
ReservedFor to the pod which triggered delayed allocation. However, this
is neither required nor enforced. Therefore we should also test the fallback
path were kube-scheduler itself adds the pod to ReservedFor.

pohly committed:
dra: revamp event handlers in kube-controller-manager

Enabling logging is useful to track what the code is doing.

There are some functional changes:
- The pod handler checks for existence of claims. This
  avoids adding pods to the work queue in more cases
  when nothing needs to be done, at the cost of
  making the event handlers a bit slower. This will become
  more important when adding more work to the controller
- The handler for deleted ResourceClaim did not check for
  cache.DeletedFinalStateUnknown.

pohly committed:
dra: handle scheduled pods in kube-controller-manager

When someone decides that a Pod should definitely run on a specific node, they
can create the Pod with spec.nodeName already set. Some custom scheduler might
do that. Then kubelet starts to check the pod and (if DRA is enabled) will
refuse to run it, either because the claims are still waiting for the first
consumer or the pod wasn't added to reservedFor. Both are things the scheduler
normally does.

Also, if a pod got scheduled while the DRA feature was off in the
kube-scheduler, a pod can reach the same state.

The resource claim controller can handle these two cases by taking over for the
kube-scheduler when nodeName is set. Triggering an allocation is simpler than
in the scheduler because all it takes is creating the right
PodSchedulingContext with spec.selectedNode set. There's no need to list nodes
because that choice was already made, permanently. Adding the pod to
reservedFor also isn't hard.

What's currently missing is triggering de-allocation of claims to re-allocate
them for the desired node. This is not important for claims that get created
for the pod from a template and then only get used once, but it might be
worthwhile to add de-allocation in the future.

pohly commented:
/cc @moshe010

cici37 commented:
/remove-sig api-machinery

bart0sh commented:
/triage accepted
/priority important-longterm
/cc @klueska

pohly commented:
/retest

pohly commented:
@alculquicondor: you were interested in how DRA would handle delayed allocation and reservation in scenarios where the normal scheduler isn't involved. The idea was to let the resource claim controller take over for those pods which already have a node name. This PR implements that. Perhaps you can review?

alculquicondor commented:
Can you point where in the KEP this mode is described?

My questions (without looking at the code) would be:
- what happens to the pod while the controller is working? I suppose it stays in Pending. Which conditions does kubelet set?
- what happens if the controller fails to provision the resource? Is the Pod marked as Failed. Does the controller set any conditions?

In general, how can a user determine if the process failed and set retry policies, if they are using a Job https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy

pohly commented:
> Can you point where in the KEP this mode is described?

It's not in the KEP yet. It seemed simple enough that a design review could get combined with the code review. I'll update the KEP once we agree that this makes sense, and we can put this PR on hold until the KEP update is also ready.

> what happens to the pod while the controller is working? I suppose it stays in Pending. 

Correct.

> Which conditions does kubelet set?

kubelet does the same as it always does in this situation. We haven't specifically defined particular condition for this. I was expecting some kind of event that informs the user, but when I tested it I found that it doesn't do that yet. @bart0sh created [an issue](https://github.com/kubernetes/kubernetes/issues/118468) for this.

Do you think that there should be special condition? Please comment in that issue.

> what happens if the controller fails to provision the resource? Is the Pod marked as Failed.

There is no detection of permanent failures. The controller will keep trying and kubelet will keep waiting.

> Does the controller set any conditions?

The controller has code to emit events. However, I need to double-check whether I am doing this correctly in this PR.

If you know of a good way to test "correct" event generation, then I am all ears. Ideally it should be done in an E2E or integration test, but that's also where we have no guarantee that a test client actually sees all events :cry:

alculquicondor commented:
> There is no detection of permanent failures. The controller will keep trying and kubelet will keep waiting.

I guess this is the same we do for PVs, so it should be fine.

Integration sounds like a good place to tests events. If you found issues, did you discuss it with api-machinery?

FYI on the topic of conditions: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation. I suggest you sync with the authors of the KEP.

alculquicondor commented:
> It seemed simple enough that a design review could get combined with the code review.

Generally, any change that involves Pod lifecycle is not simple enough :)
A good PR description could go a long way. But I think the KEP is the right way to discuss this.

pohly commented:
I've [asked on #sig-api-machinery](https://kubernetes.slack.com/archives/C0EG7JC6T/p1685990758660549) regarding event testing.

pohly committed:
dra controller: enhance testing

The allocation mode is relevant when clearing the reservedFor: for delayed
allocation, deallocation gets requested, for immediate allocation not. Both
should get tested.

All pre-defined claims now use delayed allocation, just as they would if
created normally.

elezar commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -161,80 +175,154 @@ func (ec *Controller) enqueuePod(logger klog.Logger, obj interface{}, deleted bo
+	logger.V(6).Info("not enqueing pod", "pod", klog.KObj(pod), "reason", reason)
+}
+
+func podDoesNotNeedClaims(pod *v1.Pod, deleted bool) string {
```
nit: Does returning `bool, string` here make sense? 

At the callsite we would have 
```
if notNeeded, reason := podDoesNotNeedClaims(); notNeeded {

}
```
which may make the intent clearer than checking for an empty string.

elezar replied:
(I see you use this for `podNeedsWork` below)

pohly replied:
Yes, I should use the same pattern also here. I'll change it.

elezar commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -340,6 +428,41 @@ func (ec *Controller) syncPod(ctx context.Context, namespace, name string) error
+			resourceclaim.CanBeReserved(claim) {
+			logger.V(5).Info("reserve claim for pod", "resourceClaim", klog.KObj(claim))
+			if err := ec.reserveForPod(ctx, pod, claim); err != nil {
+				return err
```
Does this need a log message since it breaks out of the loop early?

elezar reviewed (commented):
I took an initial pass at this. Some minor comments.

LGTM.

pohly commented:
@elezar: [updated](https://github.com/kubernetes/kubernetes/compare/8db3ff13ff9e87d93077e9de87dcb13c47055291..a27eb2449ff3c9df8cba80086c8d5e01b6d23906)

elezar commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -199,91 +213,170 @@ func (ec *Controller) enqueuePod(logger klog.Logger, obj interface{}, deleted bo
+			return false, "conflicting claim needs to be removed by user"
+		}
+
+		if pod.Spec.NodeName == "" {
```
Why are we checking this in the loop instead of once at the start of the function? If we only want to perform this check for non-zero lists of resource claims would `if pod.Spec.NodeName == "" && len(pod.Spec.ResourceClaims)` not be preferable as an early-return?

pohly replied:
There's indeed no good reason to not return early - changed.

pohly replied:
It has to come after `return true, err.Error()`, otherwise no claims get created :sweat_smile: 

Reverted...

pohly replied:
There also was a small logic bug in the original code: short-circuiting the entire function is too broad, we need to keep checking for other missing claims, i.e. continue the loop.

elezar commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -403,6 +496,49 @@ func (ec *Controller) syncPod(ctx context.Context, namespace, name string) error
 		}
 	}
 
+	if pod.Spec.NodeName == "" {
```
Question: How is this different to `podNeedsWork`? Seems as if much of the same logic is implemented (although I haven't compared the code paths in depth).

pohly replied:
The logic in both is very similar. The difference is that `podNeedsWork` gets invoked as part of an informer callback and has to make decisions more quickly and without blocking. Here we do the actual work.

elezar reviewed (commented):
Thanks for the changes @pohly.

I took another pass and have some minor questions and suggestions.

None are blockers though.

LGTM

pohly commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -199,91 +213,170 @@ func (ec *Controller) enqueuePod(logger klog.Logger, obj interface{}, deleted bo
 		return
 	}
 
+	if deleted {
```
Yes. This could have been a separate PR, I just happened to notice it while working on this.

https://github.com/kubernetes/kubernetes/pull/118209/commits/f340c02a9b07b9018c48b63f945fe49c4e289ff6:
> We don't need to remember that a pod got deleted when it had no resource claims
because the code which checks the cached UIDs only checks for pods which have
resource claims.

pohly commented on test/e2e/dra/test-driver/app/controller.go:
```diff
@@ -80,6 +81,7 @@ func NewController(clientset kubernetes.Interface, driverName string, resources
 func (c *ExampleController) Run(ctx context.Context, workers int) {
 	informerFactory := informers.NewSharedInformerFactory(c.clientset, 0 /* resync period */)
 	ctrl := controller.New(ctx, c.driverName, c, c.clientset, informerFactory)
+	ctrl.SetReservedFor(!c.resources.DontSetReservedFor)
```
Correct.

pohly commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -199,91 +213,170 @@ func (ec *Controller) enqueuePod(logger klog.Logger, obj interface{}, deleted bo
 				key := claimKeyPrefix + pod.Namespace + "/" + *claimName
-				logger.V(6).Info("pod is deleted or done, process claim", "pod", klog.KObj(pod), "key", key)
+				logger.V(6).Info("Process claim", "pod", klog.KObj(pod), "key", key, "reason", reason)
 				ec.queue.Add(claimKeyPrefix + pod.Namespace + "/" + *claimName)
```
Yes. I lost that as part of rebasing - will fix.

pohly commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -199,91 +213,170 @@ func (ec *Controller) enqueuePod(logger klog.Logger, obj interface{}, deleted bo
+		ec.queue.Add(podKeyPrefix + pod.Namespace + "/" + pod.Name)
+		return
+	}
+	logger.V(6).Info("not enqueing pod", "pod", klog.KObj(pod), "reason", reason)
```
I also double-checked verbosity: the intention was to log informer callback events at -v=6, but I wasn't consistent about that.

pohly commented:
Hmm, one of the recent changes must have broken something.... checking.

soltysh reviewed (approved):
The general controller changes lgtm, although I'd suggest adding a few test cases for errors, when I checked the test coverage for the controller all the error paths were red. It's not a blocker for this PR to land, but I'd suggest being more thorough in that area. 
/approve

klueska commented:
I didn't do a thorough review of all of the code myself, but in general the flow looks good and I trust @elezar's LGTM, knowing he did do a thorough review.

/lgtm

pohly commented:
/assign @liggitt 

For `plugin/pkg/auth/authorizer` approval.

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -213,6 +213,7 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
 				rbacv1helpers.NewRule("get", "list", "watch").Groups(legacyGroup).Resources("pods").RuleOrDie(),
 				rbacv1helpers.NewRule("update").Groups(legacyGroup).Resources("pods/finalizers").RuleOrDie(),
 				rbacv1helpers.NewRule("get", "list", "watch", "create", "delete").Groups(resourceGroup).Resources("resourceclaims").RuleOrDie(),
+				rbacv1helpers.NewRule("get", "list", "watch", "create", "update", "patch").Groups(resourceGroup).Resources("podschedulingcontexts").RuleOrDie(),
```
making sure it is intentional that this controller can create and update, but not delete?

liggitt replied:
> Deletion is handled through the pod owner reference and garbage collection

that's fine, just making sure

liggitt commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -152,13 +163,16 @@ func NewController(
 		},
 		UpdateFunc: func(old, updated interface{}) {
-			ec.onResourceClaimAddOrUpdate(logger, updated)
+			logger.V(6).Info("updated claim", "claimDiff", cmp.Diff(old, updated))
```
this diff will always run, regardless of whether the result is logged... isn't that expensive? also, cmp.Diff isn't guaranteed to not panic:

From https://pkg.go.dev/github.com/google/go-cmp/cmp:
> Its propensity towards panicking means that its unsuitable for production environments where a spurious panic may be fatal.

pohly replied:
You are right, let's better guard the call with an if check. I think at -v6 it's okay to take the risk of a panic. I've seen other log calls which dump diffs at higher verbosity levels.

liggitt replied:
> I think at -v6 it's okay to take the risk of a panic

Do we know this is a safe type to pass to cmp.Diff? Should we test that to give us confidence V(6) won't panic?

pohly replied:
I have tested that manually while developing the PR. For SIG Instrumentation, I am running a periodic job with logging at a very high log level which happens to include DRA. But automated testing in release informing jobs is missing.

I could add a dedicated unit test, but that's not worth it, therefore I have removed the usage of `cmp.Diff`: https://github.com/kubernetes/kubernetes/compare/5cb4f18791a44d23c2d1d83ba7323ec903e1597b..80ab8f0542f9ddcf4935e24f742ef2a94b204471

@liggitt Okay to merge now?

liggitt replied:
> @liggitt Okay to merge now?

yuup

liggitt commented:
/approve
for authorizer change

/hold
for resolution of unconditional cmp.Diff call. @pohly can unhold and lgtm once that is resolved

pohly commented:
Update pushed. Let's wait for pull-kubernetes-kind-dra before lifting the hold, just in case.

liggitt commented:
/lgtm
/hold cancel
